README.md
# porridge [![PyPI](https://img.shields.io/pypi/v/porridge.svg)](https://pypi.org/project/porridge) [![Build Status](https://travis-ci.org/thusoy/porridge.svg?branch=master)](https://travis-ci.org/thusoy/porridge) [![Windows build status](https://ci.appveyor.com/api/projects/status/y51rewx877d522b5/branch/master?svg=true)](https://ci.appveyor.com/project/thusoy/porridge/branch/master)

Simple, strong and standardized keyed password storage.

Keyed password storage utilizes server-side secrets to ensure passwords cannot be brute-forced offline if your database is leaked. A leak could happen through a SQL injection or a compromised backup, or a host of other sadly quite common webapp vulnerabilities.

While many password storage schemes like PBKDF2, bcrypt, scrypt and the likes will make recovering the passwords offline slow, a resourceful and patient attacker will eventually be able to recover most of them. Porridge makes this entirely impossible unless your secret is also leaked, which is often not the case for many common vulnerabilities. More details in my [blog post](https://thusoy.com/2017/resistance-era-password-storage) on the topic.

Note that utilizing porridge is not magical solution to passwords on the internet, a complete solution should still enforce at least a password policy, secure password resets, rate-limiting and U2F/2FA. Have experienced security engineers set up something for you, or use high-level libraries that take care of it for you.


Installing
----------

    $ pip install porridge

Pre-built wheels is available for Windows, Linux and macOS. Building from source requires cffi >= 1.0.0, which in turn requires libffi and python headers to compile.


Usage
-----

```python
from porridge import Porridge

porridge = Porridge('myfirstkey:myfirstsecret')

boiled_password = porridge.boil('password')
if porridge.verify('password', boiled_password):
    print('Success!')
else:
    print('Fail!')
```

This setup ensures that even if your database is leaked, your users' passwords are irrecoverable by an attacker that does not also possess the `myfirstsecret` value. Load it securely like you would other credentials like API keys, database passwords, session tokens and similar.

This shell snippet is handy to create strong secrets:

```bash
$ echo "$(date +%Y%m%d):$(openssl rand -base64 30)"
```

This string will thus grow regularly. After some time, it'll look something like

    keyid3:key3,keyid2:key2,keyid1:key1

The first key in the list (in this case, keyid3) will be used to boil new passwords.

Old keys are necessary to verify the passwords of users who haven't logged in since the secret was rotated. Secrets can be dropped from the config when no users have passwords using that id anymore, which you can tell from the `keyid` field in the boiled password.

**Note**: Create a `Porridge` instance once and re-use it across your application, it does some parameter validation and stuff on startup you'd rather not want to redo for every passord you process.


Local development
-----------------

    $ ./configure
    $ ./test

Continually running tests whenever source changes:

    $ ./tools/watch_and_run_tests.sh


Motivation
----------

I couldn't find any existing solutions that utilizes argon2's server-side secret feature, as most libraries only wrap the high-level interfaces, which sadly don't enable setting secrets.

Some guiding principles for this project:

- People should not have to configure cryptographic parameters
- UX is a security feature
- The database is not trusted, neither to keep things secret, nor to keep things sane
- Following nothing more than the quickstart should result in a very secure implementation
- Migrating from existing solutions should be easy


Maintenance
-----------

Keeping this running over an extended period requires two things:
    - Adding new secrets regularly (twice a year is probably fine), and whenever you suspect a breach
    - Using `needs_update()` to store new boils where the password was stored with old parameters

The first is to ensure that if your servers at one point is compromised, future passwords are not impacted. 

The second point ensures that every time one of your users log in, the parameters their existing boiled password is stored under are still strong and the secret current, otherwise it'll be re-stored.

The only thing you need to do this is to check `needs_update()` after a password has been verified, and to store the updated one if that's the case:

```python
from porridge import Porridge

porridge = Porridge('keyid2:key2,keyid1:key1')

password = ... # get this from the user
old_boiled_password = ... # Get this from your database

if porridge.verify(password, old_boiled_password):
    print('Success!')
    if porridge.needs_update(old_boiled_password):
        # update the password in the database
        new_boiled_password = porridge.boil(password)
        print('Storing new boiled password to database')
else:
    print('Fail!')
```

The default parameters will be bumped regularly with new releases of porridge, thus as long as you install updates this should keep everything fresh.

Note that to avoid DDoS itself, a Porridge instance will refuse to verify passwords boiled with parameters that are stronger than a given threshold of it's own parameters. This is to ensure that if you by accident try to verify a password with a time cost or memory cost in the millions, you will not have to wait for the heat death of the universe to regain control of your computer. But we also want to ensure that we can upgrade parameters gradually across a fleet of instances without some suddenly starting to fail, thus when you're increasing the cost parameters you should ensure you increase them with less than the `parameter_threshold`. The default threshold is 4, thus you'll be fine if you double parameters, but if you want to bump parameters with more than 4x you should increase `default_threshold` across your fleet first.


FAQ
---

*Q: I notice the word "hash" isn't used much by porridge, why?*
A: Because it's too easy to get stuff wrong when communicated to people who are not cryptographers, which include most of us. Experienced cryptographers do a mental translation of "hash" to "memory-hard key stretching" whenever they're in a password context, but the rest of us don't. Thus it's too easy for non-cryptographers to write password storage solutions that either store passwords in plaintext, or just use an actual "hash", leading to puppies dying left and right. Thus for porridge, passwords are "boiled". If non-cryptographers hear that they're supposed to boil passwords, any decent search engine will ensure they end up with a very robust solution. This project is named porridge, as it's one dish that requires salt and a long boil, but also avoids squatting a "password-boiler" package that makes it hard for other packages to attempt to solve the same problem. Eran Hammer has some [more thoughts on this](https://hueniverse.com/the-myth-of-descriptive-module-names-d34d5feaa273).

*Q: How do I migrate to porridge from pbkdf2/bcrypt/scrypt/plain argon2?*
A: Add a new column in your database to store the boiled passwords, add porridge and boil passwords with it in addition to your existing scheme and store them to the new column. When verifying, verify with both your existing scheme and porridge if there's anything in the new column. When you deem that few enough users haven't gotten their passwords boiled by porridge yet, drop the old password column and stop using the old scheme. The users who hasn't gotten new boiled passwords will be forced through password reset, but otherwise no one will notice any difference. If you are already using argon2 but without server-side secrets, porridge can dropped in directly. Note that the max length of a boiled password is *265* characters, but that requires using the associated data feature that is not exposed through porridge, so you're probably fine if your existing columns only allows 255 characters.

*Q: Why "porridge"?*
A: Because good porridge requires more than just salt, takes a long time to boil, and you cannot separate its ingredients  after they've been boiled. And I was hungry when starting writing this. And we need better terminology for password storage, see the first question.

*Q: Could you release wheels for platform X?*
A: All releases of Porridge use wheels built by Travis CI and AppVeyor (see `./tools/release.py` for how it's done). If you need another platform supported, like a specific version of PyPy on some platform, open a PR adding it to the build matrix, and it'll automatically be part of the next release.


Security
--------

Porridge wraps the reference implementation of [argon2](https://github.com/P-H-C/phc-winner-argon2), the winner of the Password Hashing Competition, which means it has been studied in detail by very experienced cryptanalysts.

The default parameters aims for ~1ms boiling time with 512kB memory usage, adjust these to suit your environment and your requirements. If you run on very performant hardware you should be able to sustain higher costs:

```python
porridge = Porridge('key1:secret1', time_cost=8, memory_cost=1024, parallelism=16)
```

`time_cost` gives the number of iterations of the core algorithm, measure to see the impact in your environment.
`memory_cost` is how many kibibytes of memory to use for each password boiled.
`parallelism` is how many threads to use for each password boiled.

If you find a security-critical bug that you'd rather not disclose openly in the issues, shoot an email to hello at thusoy.com. This project does not have a bug bounty, but you will be credited (if you wish) here in the README and in the changelog.


Alternatives
------------

If you can't apply server-side secrets, plain [argon2](https://github.com/hynek/argon2_cffi) is the recommended way to store your passwords as of best practices in 2017. To utilize server-side secrets with other schemes you can HMAC the password with your secret before passing it to your key stretching function, but it'll be very hard to rotate this secret without invalidating all your passwords.


Credits
-------

Many thanks to [argon2_cffi](https://github.com/hynek/argon2_cffi) for a great starting point for wrapping argon2.


Troubleshooting
---------------

### Import fails with libpython shared object not found

If you get a traceback similar to this when trying to import the module:

```
    from ._ffi import ffi, lib
ImportError: libpython2.7.so.1.0: cannot open shared object file: No such file or directory
```

You are probably missing the `python-dev` package. On ubuntu/debian: `sudo apt-get install python-dev`.


### Installation fails with 'compilation terminated'

If installing the module fails with a traceback like
```
c/_cffi_backend.c:15:17: fatal error: ffi.h: No such file or directory

 #include <ffi.h>

                 ^

compilation terminated.

error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
```

You are trying to install from source (might happen if using an older pip which doesn't install manylinux1 wheels) and the compiler can't find the libraries it needs. On ubuntu/debian: `sudo apt-get install libffi-dev python-dev`. Updating your pip should also work and might not require root if you're running in a virtualenv.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).


1.0.0 - 2017-05-04
------------------

First release!


porridge/porridge.py
import contextlib
import os
import re

from ._ffi import ffi, lib
from .utils import (
    b64_decode_raw,
    b64_encode_raw,
    check_types,
    ensure_bytes,
    string_types,
)
from .exceptions import (
    EncodedPasswordError,
    MissingKeyError,
    ParameterError,
    PorridgeError,
)

# These parameters should be increased regularly to keep boiling slow
# on new hardware
DEFAULT_RANDOM_SALT_LENGTH = 16
DEFAULT_HASH_LENGTH = 32
DEFAULT_TIME_COST = 2
DEFAULT_MEMORY_COST = 512
DEFAULT_PARALLELISM = 4
DEFAULT_PARAMETER_THRESHOLD = 4
DEFAULT_FLAGS = lib.ARGON2_FLAG_CLEAR_PASSWORD | lib.ARGON2_FLAG_CLEAR_SECRET

# This regex validates the spec from
# https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md
ENCODED_HASH_RE = re.compile(r''.join([
        r'^\$argon2i\$',
        r'(?:v=(?P<version>[0-9]{1,3})\$)?',
        r''.join([
            r'm=(?P<memory_cost>[0-9]{1,10})',
            r',t=(?P<time_cost>[0-9]{1,10})',
            r',p=(?P<parallelism>[0-9]{1,3})',
            r'(?:,keyid=(?P<keyid>[a-zA-Z0-9+/]{0,11}))?', # optional
            r'(?:,data=(?P<data>[a-zA-Z0-9+/]{0,43}))?', # optional, unused
        ]),
        r'\$(?P<salt>[a-zA-Z0-9+/]{11,64})\$',
        r'(?P<hash>[a-zA-Z0-9+/]{16,86})',
    ]) + r'$'
)


class Porridge(object):
    r"""
    Helper class to boil passwords with sensible defaults and server-side
    secrets.

    :param str secrets: A comma-separated string of *keyid:key* pairs that will
        be used as server-side secrets. The first element in the list will be
        used to boil new passwords, the others to verify old ones.
    :param int time_cost: Defines the amount of computation realized and
        therefore the execution time, given in number of iterations.
    :param int memory_cost: Defines the memory usage, given in kibibytes_.
    :param int parallelism: Defines the number of threads used
    :param int hash_len: Length of the raw hash in bytes.
    :param int salt_len: Length of random salt to be generated for each
        password.
    :param int parameter_threshold: A multiplier that sets the threshold for
        how much higher parameters in boiled passwords can be above our own
        parameters before we refuse the process them.
    :param str encoding: Boiling is always performed on bytes, thus if unicode
        strings are given to either :meth:`boil` of :meth:`verify` this encoding
        will be used to encode the password to bytes.

    .. _salt: https://en.wikipedia.org/wiki/Salt_(cryptography)
    .. _kibibytes: https://en.wikipedia.org/wiki/Binary_prefix#kibi
    """

    def __init__(
        self,
        secrets,
        time_cost=DEFAULT_TIME_COST,
        memory_cost=DEFAULT_MEMORY_COST,
        parallelism=DEFAULT_PARALLELISM,
        hash_len=DEFAULT_HASH_LENGTH,
        salt_len=DEFAULT_RANDOM_SALT_LENGTH,
        parameter_threshold=DEFAULT_PARAMETER_THRESHOLD,
        encoding="utf-8",
    ):
        e = check_types(
            secrets=(secrets, string_types),
            time_cost=(time_cost, int),
            memory_cost=(memory_cost, int),
            parallelism=(parallelism, int),
            hash_len=(hash_len, int),
            salt_len=(salt_len, int),
            parameter_threshold=(parameter_threshold, int),
            encoding=(encoding, string_types),
        )
        if e:
            raise TypeError(e)

        self.time_cost = time_cost
        self.memory_cost = memory_cost
        self.parallelism = parallelism
        self.hash_len = hash_len
        self.salt_len = salt_len
        self.encoding = encoding
        if parameter_threshold < 1:
            raise ValueError('parameter_threshold must be at least 1')
        self.parameter_threshold = parameter_threshold

        self.secret_map = {}
        self.secret = None
        self.keyid = None
        for secret_pair in secrets.split(','):
            keyid, secret = secret_pair.split(':', 1)
            keyid = keyid.encode('utf-8')
            secret = secret.encode('utf-8')
            if self.secret is None:
                self.secret = secret
                self.keyid = keyid
            self.secret_map[keyid] = secret

        self._self_check()


    def _self_check(self):
        """
        Perform a single run of boiling to ensure we have a valid
        combination of parameters.
        """
        self.boil('dummy')


    def boil(self, password):
        """
        Boil *password* and return and boiled password that can be stored in a
        database.

        :param password: Password to boil.
        :type password: ``bytes`` or ``unicode``

        :raises porridge.PorridgeError: If verification fails to complete due
            to not being able to spawn enough threads or allocate enough memory.

        :rtype: unicode
        """
        e = check_types(password=(password, string_types + (bytes,)))
        if e:
            raise TypeError(e)

        salt = os.urandom(self.salt_len)
        context_params = dict(
            salt=salt,
            password=ensure_bytes(password, self.encoding),
            secret=self.secret,
            time_cost=self.time_cost,
            memory_cost=self.memory_cost,
            parallelism=self.parallelism,
            hash_len=self.hash_len,
        )
        with argon2_context(**context_params) as ctx:
            result = compute_hash(ctx)

            if result != lib.ARGON2_OK:
                error_message = argon2_error_message(result)
                if is_operational_error(result):
                    raise PorridgeError(error_message)
                else:
                    raise ParameterError(error_message)

            raw_hash = bytes(ffi.buffer(ctx.out, ctx.outlen))
        return self._encode(raw_hash, salt)


    def verify(self, password, boiled):
        """
        Verify that *password* matches *boiled*.

        :param password: The password to verify.
        :type password: ``bytes`` or ``unicode``
        :param unicode boiled: An boiled password as returned from
            :meth:`Porridge.boil`.

        :raises porridge.PorridgeError: If verification fails to complete due
            to not being able to spawn enough threads or allocate enough memory.

        :return: ``True`` if *password* is valid, otherwise ``False``.
        :rtype: bool
        """
        e = check_types(
            password=(password, string_types + (bytes,)),
            boiled=(boiled, string_types),
        )
        if e:
            raise TypeError(e)

        if len(boiled) > 265:
             # Ensure we don't DDoS ourselves if the database holds corrupt values
            raise EncodedPasswordError('Encoded password exceeds maximum length of '
                '265, was {length}'.format(length=len(boiled)))

        context_params = parse_boiled(boiled)
        self._verify_parameters_within_threshold(context_params)
        raw_hash = context_params.pop('raw_hash')

        context_params.update(dict(
            hash_len=len(raw_hash),
            password=self._ensure_bytes(password),
        ))

        keyid = context_params.get('keyid')
        if keyid:
            del context_params['keyid']
            secret = self.secret_map.get(keyid)
            if not secret:
                raise MissingKeyError(keyid.decode('utf-8'))
            context_params['secret'] = secret

        with argon2_context(**context_params) as ctx:
            result = verify_hash(ctx, raw_hash)

        if result == lib.ARGON2_OK:
            return True
        elif result == lib.ARGON2_VERIFY_MISMATCH:
            return False
        else:
            error_message = argon2_error_message(result)
            raise PorridgeError(error_message)


    def needs_update(self, boiled):
        """
        Check if the parameters in *boiled* are old and the password should be
        re-boiled.

        :param unicode boiled: An boiled password as returned from
            :meth:`Porridge.boil`.

        :rtype: bool
        """
        parsed = parse_boiled(boiled)
        if parsed['version'] < lib.ARGON2_VERSION_NUMBER:
            return True

        if parsed['parallelism'] < self.parallelism:
            return True

        if parsed['memory_cost'] < self.memory_cost:
            return True

        if parsed['time_cost'] < self.time_cost:
            return True

        if len(parsed['salt']) < self.salt_len:
            return True

        if len(parsed['raw_hash']) < self.hash_len:
            return True

        if parsed.get('keyid') != self.keyid:
            return True

        return False


    def _ensure_bytes(self, s):
        return ensure_bytes(s, self.encoding)


    def _verify_parameters_within_threshold(self, parameters):
        for parameter in ('time_cost', 'memory_cost', 'parallelism'):
            given_parameter = parameters[parameter]
            our_parameter = getattr(self, parameter)
            if given_parameter > our_parameter * self.parameter_threshold:
                raise EncodedPasswordError('%s exceeds threshold of what we will process' % parameter)


    def _encode(self, raw_hash, salt):
        template = (
            '${algo}$v={version}$m={m_cost},t={t_cost},p={parallelism}'
            ',keyid={keyid}${salt}${hash}'
        )
        return template.format(
                algo='argon2i',
                t_cost=self.time_cost,
                m_cost=self.memory_cost,
                parallelism=self.parallelism,
                salt=b64_encode_raw(salt),
                hash=b64_encode_raw(raw_hash),
                version=lib.ARGON2_VERSION_NUMBER,
                keyid=self.keyid.decode('utf-8'),
            )


    def __str__(self):
        return ("Porridge(key='{key}', memory_cost={memory_cost}, "
            "time_cost={time_cost}, parallelism={parallelism})").format(
            key=self.keyid.decode('utf-8'),
            memory_cost=self.memory_cost,
            time_cost=self.time_cost,
            parallelism=self.parallelism,
        )


    def __repr__(self):
        return ("Porridge(key='{key}', memory_cost={memory_cost}, time_cost={time_cost}, "
            "parallelism={parallelism}, hash_len={hash_len}, salt_len={salt_len}, "
            "parameter_threshold={parameter_threshold}, encoding='{encoding}')").format(
            key=self.keyid.decode('utf-8'),
            memory_cost=self.memory_cost,
            time_cost=self.time_cost,
            parallelism=self.parallelism,
            hash_len=self.hash_len,
            salt_len=self.salt_len,
            parameter_threshold=self.parameter_threshold,
            encoding=self.encoding,
        )


def parse_boiled(boiled):
    match = ENCODED_HASH_RE.match(boiled)
    if not match:
        raise EncodedPasswordError('Encoded password is on unknown format', boiled)
    version = match.group('version')
    if version:
        version = int(version)
    else:
        # Default to the old version as only ARGON2_VERSION_13 includes it in the boiled string
        version = lib.ARGON2_VERSION_10

    salt = b64_decode_raw(match.group('salt'))
    raw_hash = b64_decode_raw(match.group('hash'))
    time_cost = int(match.group('time_cost'))
    memory_cost = int(match.group('memory_cost'))
    parallelism = int(match.group('parallelism'))

    parsed = dict(
        time_cost=time_cost,
        memory_cost=memory_cost,
        parallelism=parallelism,
        raw_hash=raw_hash,
        salt=salt,
        version=version,
    )

    keyid = match.group('keyid')
    if keyid:
        parsed['keyid'] = keyid.encode('utf-8')

    return parsed


def argon2_error_message(error_code):
    return ffi.string(lib.argon2_error_message(error_code)).decode('utf-8')


def is_operational_error(error_code):
    return error_code in set([
        lib.ARGON2_THREAD_FAIL,
        lib.ARGON2_MEMORY_ALLOCATION_ERROR,
    ])


def compute_hash(context):
    '''Minimal wrapper around argon2_ctx to enable mocking'''
    return lib.argon2_ctx(context, lib.Argon2_i)


def verify_hash(context, raw_hash):
    '''Minimal wrapper around argon2i_verify_ctx to enable mocking'''
    return lib.argon2i_verify_ctx(context, raw_hash)


@contextlib.contextmanager
def argon2_context(
        password=None, # bytes
        salt=None,
        secret=None,
        hash_len=DEFAULT_HASH_LENGTH,
        time_cost=DEFAULT_TIME_COST,
        memory_cost=DEFAULT_MEMORY_COST,
        parallelism=DEFAULT_PARALLELISM,
        flags=DEFAULT_FLAGS,
        version=lib.ARGON2_VERSION_NUMBER,
        ):
    csalt = ffi.new("uint8_t[]", salt)
    cout = ffi.new("uint8_t[]", hash_len)
    cpwd = ffi.new("uint8_t[]", password)

    if secret:
        csecret = ffi.new("uint8_t[]", secret)
        secret_len = len(secret)
    else:
        csecret = ffi.NULL
        secret_len = 0

    ctx = ffi.new("argon2_context *", dict(
            version=version,
            out=cout, outlen=hash_len,
            pwd=cpwd, pwdlen=len(password),
            salt=csalt, saltlen=len(salt),
            secret=csecret, secretlen=secret_len,
            ad=ffi.NULL, adlen=0,
            t_cost=time_cost,
            m_cost=memory_cost,
            lanes=parallelism, threads=parallelism,
            allocate_cbk=ffi.NULL, free_cbk=ffi.NULL,
            flags=flags,
        )
    )
    yield ctx
